src/riak_cs_gc_batch.erl

```diff
@@ -0,0 +1,313 @@
+    %%?debugFmt("StateName ~p, ~p ~w", [StateName, has_batch_finished(State), State]),
+    case {has_batch_finished(State), StateName} of
+        {true, _} ->
+            {stop, normal, State};
+        {false, waiting_for_workers} ->
+            try_next_batch(State)
+    end;
+handle_event(_Event, StateName, State) ->
+    {next_state, StateName, State}.
+
+%% @doc Handle synchronous events that should be handled
+%% the same regardless of the current state.
+-spec handle_sync_event(term(), term(), atom(), ?STATE{}) ->
+                               {reply, term(), atom(), ?STATE{}}.
+handle_sync_event(current_state, _From, StateName, State) ->
+    {reply, {StateName, State}, StateName, State};
+handle_sync_event(stop, _From, _StateName, State) ->
+    _ = cancel_batch(State),
+    {stop, cancel, {ok, State}, State};
+handle_sync_event(_Event, _From, StateName, State) ->
+    ok_reply(StateName, State).
+
+handle_info(_Info, StateName, State) ->
+    {next_state, StateName, State}.
+
+%% @doc TODO: log warnings if this fsm is asked to terminate in the
+%% middle of running a gc batch
+terminate(normal, _StateName, State) ->
+    _ = lager:info("Finished garbage collection: "
+                   "~b seconds, ~p batch_count, ~p batch_skips, "
+                   "~p manif_count, ~p block_count\n",
+                   [elapsed(State?STATE.batch_start), State?STATE.batch_count,
+                    State?STATE.batch_skips, State?STATE.manif_count,
+                    State?STATE.block_count]),
+    riak_cs_gc_manager:finished(State);
+terminate(_Reason, _StateName, _State) ->
+    ok.
+
+%% @doc this fsm has no special upgrade process
+code_change(_OldVsn, StateName, State, _Extra) ->
+    {ok, StateName, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+maybe_fetch_first_key(?STATE{batch_start=BatchStart,
+                             leeway=Leeway} = State) ->
+
+    %% [Fetch the first set of manifests for deletion]
+    %% this does not check out a worker from the riak connection pool;
+    %% instead it creates a fresh new worker, the idea being that we
+    %% don't want to delay deletion just because the normal request
+    %% pool is empty; pool workers just happen to be literally the
+    %% socket process, so "starting" one here is the same as opening a
+    %% connection, and avoids duplicating the configuration lookup
+    %% code.
+    {KeyListRes, KeyListState} =
+        riak_cs_gc_key_list:new(BatchStart, Leeway),
+    #gc_key_list_result{bag_id=BagId, batch=Batch} = KeyListRes,
+    _ = lager:debug("Initial batch keys: ~p", [Batch]),
+    State?STATE{batch=Batch,
+                key_list_state=KeyListState,
+                bag_id=BagId}.
+
+maybe_fetch_next_keys(?STATE{key_list_state=undefined} = State) ->
+    State;
+maybe_fetch_next_keys(?STATE{key_list_state=KeyListState} = State) ->
```

Comment:
If `key_list_state` is not `undefined`, it should be checked there is more results before issuing 2i fetch.
The cost of 2i fetch is rather high (at least much higher than on memory checking of record fields).